some fixes + implemented AFC

mister-r opened the pull request:
Fixed several issues: race condition on initialization, RPI compilations with FFT sizes different  than 512.
Implemented Automatic Frequency Control (controlled by config, off by default). However when enable AFC works only on moment when squelch goes from ' ' to '*' by checking if some nearest freq has higher level. And moment when squelch goes back to ' ' - channel frequency reset back to config-defined value.
Added  -ffast-math option to GCC command line that allowed hardware FPU/NEON support. This subsequently allowed to get rid of asm-coded fftwave without performance degradation (actually it even a bit faster now). Unfortunatelly samplefft written in asm still bit faster than C code so it remain (but fixed to support different FFT sizes)

szpajder commented:
Thanks, this is good stuff. A few comments:

On RPi v1 we don't have NEON onboard:

    g++ -O3 -g -Wall -DSYSCONFDIR=\"/usr/local/etc\" -DUSE_BCM_VC -I/opt/vc/include  -I/opt/vc/include/interface/vcos/pthreads -I/opt/vc/include/interface/vmcs_host/linux -mcpu=arm1176jzf-s -mtune=arm1176jzf-s -march=armv6zk -mfpu=vfp -ffast-math    -c -o rtl_airband.o rtl_airband.cpp
    In file included from rtl_airband.cpp:126:0:
    /usr/lib/gcc/arm-linux-gnueabihf/4.6/include/arm_neon.h:32:2: error: #error You must enable NEON instructions (e.g. -mfloat-abi=softfp -mfpu=neon) to use arm_neon.h
    <builtin>: recipe for target 'rtl_airband.o' failed
    make: *** [rtl_airband.o] Error 1

You don't seem to use any intrinsics, so `arm_neon.h` is probably not necessary? The program compiles and works without it.

Please make AFC less verbose. It prints the log message every time the squelch opens. If syslog logging is enabled, it would clog the log files pretty quickly.

Could you please resubmit this request against unstable branch? This is the place for new features and major code changes. Master branch is for releases only.

Thanks again for your effort.

mister-r commented:
Ah, sorry arm_neon.h remained from experiments: I tried to implement sample_fft using NEON intristics but it still was bit slower than pure asm.
Will check unstable branch bit later..

mister-r commented:
I also added 'append' option. Its useful in the case when stick is used not full time for monitoring, but sometimes I want to 'grab' it temporary for other purposes and re-start monitoring when not need it anymore. Currently this causes mp3 files to be lost. Its possible to backup that files before restart rtl_airband but the time of sounds on record will be messed. So I added this 'append' option that on file opening - writes silence of period during which file was not used. And silence bounds are marked with special tones to know that this silence is due to outage of rtl_airband, but not due to actual air silence..
